src/mono/System.Private.CoreLib/src/System/Reflection/Emit/DerivedTypes.Mono.cs

```diff
@@ -1,502 +0,0 @@
-        public override Type GetNestedType(string name, BindingFlags bindingAttr)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-
-        [DynamicallyAccessedMembers(GetAllMembers)]
-        public override MemberInfo[] GetMember(string name, MemberTypes type, BindingFlags bindingAttr)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-
-       [DynamicallyAccessedMembers(GetAllMembers)]
-        public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-
-        public override InterfaceMapping GetInterfaceMap([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)] Type interfaceType)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-
-        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicEvents | DynamicallyAccessedMemberTypes.NonPublicEvents)]
-        public override EventInfo[] GetEvents(BindingFlags bindingAttr)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-
-        protected override TypeAttributes GetAttributeFlagsImpl()
-        {
-            // Return the attribute flags of the base type?
-            Type baseType;
-            for (baseType = m_baseType; baseType is SymbolType; baseType = ((SymbolType)baseType).m_baseType) ;
-            return baseType.Attributes;
-        }
-
-        protected override bool IsPrimitiveImpl()
-        {
-            return false;
-        }
-
-        protected override bool IsValueTypeImpl()
-        {
-            return false;
-        }
-
-        protected override bool IsCOMObjectImpl()
-        {
-            return false;
-        }
-
-        public override bool IsConstructedGenericType
-        {
-            get
-            {
-                return false;
-            }
-        }
-
-        public override Type GetElementType()
-        {
-            return m_baseType;
-        }
-
-        protected override bool HasElementTypeImpl()
-        {
-            return m_baseType != null;
-        }
-
-        public override object[] GetCustomAttributes(bool inherit)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-
-        public override object[] GetCustomAttributes(Type attributeType, bool inherit)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-
-        public override bool IsDefined(Type attributeType, bool inherit)
-        {
-            throw new NotSupportedException(SR.NotSupported_NonReflectedType);
-        }
-    }
-
-    [StructLayout(LayoutKind.Sequential)]
-    internal sealed class ArrayType : SymbolType
```

Comment:
These types removed and no test failures found, but there are some references left in mono native code:
https://github.com/dotnet/runtime/blob/59dddfdef247b9b85a69d663a1d1c1746206711e/src/mono/mono/metadata/sre.c#L1712-L1747

It looks to me like they are not used/needed and all these references can be removed (other references are just declarations). @lambdageek @marek-safar @vargaz please let me know what you think or with what scenarios I could test this functionality.

Reply:
In theory, this code can get executed in reflection.emit scenarios when somebody creates a pointer type from a type builder and maybe that type is referenced by code generated using ILGenerator etc.

Reply:
@buyaa-n yea I agree, these are still needed.

I think something like this should exercise that code:

```csharp
TypeBuilder tb = modBuilder.DefineType("SomeType"); // define a new type

Type ptrToTB = tb.MakePointerType();  // This should return a Sytem.Reflection.Emit.PointerType instance

ILGenerator ilg = ... ; // add some method and start adding IL to it

// emit "return typeof(SomeType*).FullName"
ilg.Emit(OpCodes.Ldtoken, ptrToTB);
ilg.Emit(OpCodes.Call, typeof(System.Type).GetMethod("GetTypeFromHandle", BindingFlags.Static));
ilg.Emit(OpCodes.Callvirt, typeof(System.Type).GetMethod("get_FullName"));
ilg.Emit(OpCodes.Ret);

var t = tb.CreateType();

var mi = ... .GetMethod(...); // get the method that ilg was building
string s = (string) mi.Invoke(...);  // get back "SomeType*"
```

---

That said, I think the CoreCLR approach of having a single `SymbolType` with a `m_typeKind` field to differentiate between arrays, pointers and byrefs is reasonable too.  We would just need a `MonoReflectionSymbolType` struct defined in `object-internals.h` and then use it in place of `MonoReflectionDerivedType` and `MonoReflectionArrayType`; as well as updte the definitions of `is_sre_array`, `is_sre_pointer`, `is_sre_byref` etc.

Reply:
Probably need a SymbolType.Mono.cs file as well with the fields required the mono runtime code.

Reply:
Thanks for the response, I will give it a try in few days.

Reply:
I added test cases hitting that path and it passes on CoreClr but fails on Mono:
https://github.com/dotnet/runtime/blob/fc0c8075e792e3ce7520c81c5127db6ad0bd9389/src/libraries/System.Reflection.Emit/tests/MethodBuilder/MethodBuilderGetILGenerator.cs#L86-L103

Updated the mono code to use MonoReflectionSymbolType in place of MonoReflectionDerivedType and MonoReflectionArrayType with related changes but the tests fails with:  
```log
 * Assertion at D:\dotnet\runtime\src\mono\mono\metadata\class.c:4700, condition `is_ok (error)' not met, function:mono_ldtoken_checked, (null) assembly:D:\dotnet\runtime\artifacts\bin\testhost\
  net8.0-windows-Debug-x64\shared\Microsoft.NETCore.App\8.0.0\System.Private.CoreLib.dll type:NotImplementedException member:(null)
```
It appears to me that it did not hit the native code path I updated:
https://github.com/dotnet/runtime/blob/fc0c8075e792e3ce7520c81c5127db6ad0bd9389/src/mono/mono/metadata/sre.c#L1698-L1705
It fails somewhere before that path when try to load a SymbolType and could not figure out what is going on

Reply:
> It fails somewhere before that path

Without debugging (probably in this case using `Internal.Console.WriteLine`), I can only guess.  From the assertion at class.c:4700, I think `ModuleBuilder.GetPseudoToken` isn't being called for the `SymbolType` (or more precisely `GetPseudoToken` isn't calling `RegisterToken` for some reason).

(Locally I tried your new added test with mono on the `main` branch, and it gives the expected result)

Reply:
@buyaa-n when I try your branch locally, looks like `SymbolType` is missing `RuntimeResolve` and `InternalResolve` methods that mono's old `ArrayType`, `ByRefType` and `PointerType` used to have.  I think those need to be added back.

Reply:
Thank you, that makes it work, except `ByRefType` and I don't see any code difference from removed code. The `ByRefType` test fails even without these changes, and it passes for CoreClr, so I assume there is existing bug and planning to create an issue for that.